### PR TITLE
Driver methods

### LINKED ISSUE #864: Add driver functions for gtclang, dawn-opt, and dawn-codegen

### PR DESCRIPTION
**Pull request by jdahm**
## Technical Description

Add drivers for each gtclang and dawn component and uses these in the executables. This does add code, but that is mostly in separating options, and more code will be eliminated once only these drivers are used (follow-up PR).

### Resolves / Enhances

Resolves #864.

### CODE REVIEW

File: dawn/src/dawn/CodeGen/Options.inc [hunk 1]
```diff
@@ -32,6 +32,11 @@
 //                  command-line target will be generated which allows to negate the option i.e
 //
 
+// TODO Rename this to (MaxHaloPoints, "max-halo") after combined struct is eliminated
```
Comment (jdahm) [inline dawn/src/dawn/CodeGen/Options.inc]: This TODO is somewhat important and will be eventually removed, so I vote we keep it for now.

**Comment by jdahm** [comment]
launch jenkins

**Comment by jdahm** [comment]
launch jenkins

**Comment by jdahm** [comment]
launch jenkins

**Comment by jdahm** [comment]
daint crashed during this build. Trying again...

**Comment by jdahm** [comment]
launch jenkins

**Comment by jdahm** [comment]
launch jenkins

**Comment by jdahm** [comment]
another daint error. Trying again...

**Comment by jdahm** [comment]
launch jenkins

**Comment by jdahm** [comment]
@twicki This is ready for review!

File: dawn/src/dawn/CodeGen/Options.inc (hunk 1 above)
Comment (twicki) [reply dawn/src/dawn/CodeGen/Options.inc]: :+1:

File: dawn/src/dawn/Compiler/DawnCompiler.h [hunk 2]
```diff
@@ -15,37 +15,22 @@
 #ifndef DAWN_COMPILER_DAWNCOMPILER_H
 #define DAWN_COMPILER_DAWNCOMPILER_H
 
+#include "dawn/CodeGen/Options.h"
```
Comment (twicki) [inline dawn/src/dawn/Compiler/DawnCompiler.h]: why di we have more includes here if we only removed code?

File: dawn/src/dawn/Compiler/Driver.cpp [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+//===------------------------------------------------------------------------------------------===//
+
+#include "dawn/Compiler/Driver.h"
+#include "dawn/CodeGen/TranslationUnit.h"
+#include "dawn/Compiler/DawnCompiler.h"
+#include "dawn/Compiler/Options.h"
+
+namespace dawn {
+
+std::list<PassGroup> defaultPassGroups() {
```
Comment (twicki) [inline dawn/src/dawn/Compiler/Driver.cpp]: should this live here? Would this not be nicer in the optimizer?

File: gtclang/src/gtclang/Driver/Driver.h [hunk 4]
```diff
@@ -19,7 +19,9 @@
 
 #include "dawn/SIR/SIR.h"
 #include "dawn/Support/NonCopyable.h"
+#include "gtclang/Driver/Options.h"
 #include "llvm/ADT/SmallVector.h"
+#include <list>
```
Comment (twicki) [inline gtclang/src/gtclang/Driver/Driver.h]: list seems unnecessary here

File: dawn/src/dawn/CodeGen/Cuda/CudaCodeGen.cpp [hunk 5]
```diff
@@ -52,7 +52,24 @@ std::string makeIntervalBoundExplicit(std::string dim, const iir::Interval& inte
...
 } // namespace
 
-CudaCodeGen::CudaCodeGen(const stencilInstantiationContext& ctx, DiagnosticsEngine& engine,
+std::unique_ptr<TranslationUnit>
+run(const std::map<std::string, std::shared_ptr<iir::StencilInstantiation>>&
+        stencilInstantiationMap,
+    const Options& options) {
+  DiagnosticsEngine diagnostics;
+  const Array3i domain_size{options.DomainSizeI, options.DomainSizeJ, options.DomainSizeK};
+  CudaCodeGen CG(stencilInstantiationMap, diagnostics, options.MaxHaloSize, options.nsms,
```
[the remark this answers is not in the document]
Comment (jdahm) [reply dawn/src/dawn/CodeGen/Cuda/CudaCodeGen.cpp]: Yes! But I would like to limit the scope of the PR to adding driver methods, then we can massage the inner workings of the drivers in another PR. As you aptly point out yes these objects can now be altered to take all options.

File: dawn/src/dawn/Compiler/DawnCompiler.h (hunk 2 above)
Comment (jdahm) [reply dawn/src/dawn/Compiler/DawnCompiler.h]: Taking a look now.

File: dawn/src/dawn/CodeGen/Driver.h [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+namespace dawn {
+namespace codegen {
+
+/// @brief Parse the backend string to enumeration
+Backend parseBackendString(const std::string& backendStr);
+
+/// @brief Run the code generation
+std::unique_ptr<TranslationUnit>
+run(const std::map<std::string, std::shared_ptr<iir::StencilInstantiation>>& context,
```
[the remark this answers is not in the document]
Comment (jdahm) [reply dawn/src/dawn/CodeGen/Driver.h:33]: Yes! In next PR.

File: dawn/src/dawn/Compiler/Driver.cpp (hunk 3 above)
Comment (jdahm) [reply dawn/src/dawn/Compiler/Driver.cpp:22]: The next PR does this - #886. In this PR it should be in Optimizer/OptimizerOptions, but that doesn't have a corresponding source file yet.

File: gtclang/src/gtclang/Driver/Driver.h (hunk 4 above)
Comment (jdahm) [reply gtclang/src/gtclang/Driver/Driver.h]: Good call!

File: dawn/src/dawn/Compiler/DawnCompiler.h (hunk 2 above)
Comment (jdahm) [reply dawn/src/dawn/Compiler/DawnCompiler.h]: Good call!

**Comment by jdahm** [comment]
launch jenkins

**Comment by jdahm** [comment]
launch jenkins

**Comment by jdahm** [comment]
Is this ready then @twicki?

**Pull request merged**